File: java/0006_ZigZag_Conversion.java
class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1) {
            return s;
        }

        StringBuilder sb = new StringBuilder();
        int len = s.length();

        for (int i = 0; i < numRows; i++) {
            boolean fromBelow = true;
            for (int j = i; j < len;) {
                sb.append(s.charAt(j));

                int backFromBelow = 2 * ((numRows - 1 ) - i);
                int backFromAbove = 2 * i;

                if (fromBelow) {
                    j += (backFromBelow > 0) ? backFromBelow : backFromAbove;
                    fromBelow = false;
                } else {
                    j += (backFromAbove > 0) ? backFromAbove : backFromBelow;
                    fromBelow = true;
                }
            }
        }

        return sb.toString();
    }
}

File: kotlin/0006_ZigZag_Conversion.kt
class Solution {
    fun convert(s: String, numRows: Int): String {
        if (numRows == 1) {
            return s
        }

        val sb = StringBuilder()
        val len = s.length
        val lastRow = numRows - 1
        val longest = 2 * lastRow

        for (i in 0 until numRows) {
            var fromBelow = true
            var j = i
            while (j < len) {
                sb.append(s[j])

                if (i in 1 until lastRow) {
                    j += if (fromBelow) 2 * (lastRow - i) else 2 * i
                    fromBelow = !fromBelow
                } else {
                    j += longest
                }
            }
        }
        return sb.toString()
    }
}